Elasticsearch: Async handling of indexing/deletion requests

tobias-hotz opened the pull request:
Currently, indexing is batched as a global queue of 200 elements (except when forceRefreshReaders is true). When the threshold is reached, the entries are submitted, and the thread submitting the 200th element waits until the elasticsearch returns the result of the request.
With deletion, we currently always send one request per deletion request and always use the deleteByQuery method.

The current design has a number of flaws:
- When multiple threads index metadata and both do not use forceRefreshReaders, the queue could grow past 200 (e.g., two threads add at about the same time, and at the time the queue size is checked, the queue already has 201 elements). This causes the queue to grow indefinitely, meaning no further entries are submitted before an explicit call to sendDocumentsToIndex. This can cause Out Of Memory errors (we observed this when using multiple reindexing threads)
- An indexing thread is spending a significant amount of time waiting for the elasticsearch, while it could already prepare the next batch for the ES to be consumed
- As we wait after every delete call, deleting many entries takes a lot more time as needed
- Due to deletion always using deleteByQuery (even when we could delete by uuid), no batching is possible
- When multiple threads index metadata and both do not use forceRefreshReaders, the queue contains mixed entries of Thread 1 and Thread 2. At most call sites, after all entries have been indexed, sendDocumentsToIndex is called. It is currently possible that Thread 1 is currently sending a batch that contains entries from Thread 2 as well and Thread 2 already finished submitting the rest of the batch, causing Thread 2 to think that all its entries have been sent to Elasticsearch, but they are still being submitted by Thread 1. This is a very rare problem, though

This PR solves all of these problems. The main takeaway is that it significantly improves the performance of deleting and indexing many entries.

This is accomplished by introducing a `IIndexSubmitter` and `IDeletionSubmitter`. These new classes handle how new entries are sent to the index. The direct implementations (`DirectIndexSubmitter` and `DirectDeletionSubmitter`) are similar to how the old `forceIndexChanges` parameter worked in that they directly send the data to the index.
With the use of the `BatchingIndexSubmitter` and `BatchingDeletionSubmittor`, chunks are sent periodically to the elasticsearch (just as before), but a local queue is used, and we do not wait for the elasticearch. The index responses are handled asynchronously on a different thread instead. We still guarantee that the indexing will be complete once the whole block is done, as the close method sends the rest of the local queue and waits for all async responses to be complete.

We made some performance measurements on a smaller scale. Here is the average result of a bunch of runs with different CSW harvesters:

| Operation                | Average time before change | Average time after change | Improvement in % |
| ------------------------ | -------------------------- | ------------------------- | ---------------- |
| Harvesting ~3600 entries | 11 minutes and 17 seconds  | 06 minutes and 30 seconds | ~74%             |
| Harvesting ~1250 entries | 06 minutes and 33 seconds  | 04 minutes and 28 seconds | ~47%             |
| Harvesting ~700 entries  | 03 minutes and 14 seconds  | 01 minutes and 30 seconds | ~131%            |
| Harvesting ~100 entries  | 00 minutes and 29 seconds  | 00 minutes and 29 seconds | ~35%             |
| Reindexing ~5700 entries | 04 minutes and 39 seconds  | 02 minutes and 40 seconds | ~74%             |
| Deleting ~3600 entries   | 01 minutes and 35 seconds  | 00 minutes and 20 seconds | ~475%            |
| Deleting ~1250 entries   | 00 minutes and 38 seconds  | 00 minutes and 09 seconds | ~433%            |
| Deleting ~700 entries    | 00 minutes and 20 seconds  | 00 minutes and 04 seconds | ~500%            |
| Deleting ~100 entries    | 00 minutes and 05 seconds  | 00 minutes and 01 seconds | ~500%            |


As you can see, there are very significant performance gains. These numbers were recorded on a local machine, if you use a remote index on a different machine, the effect may be even higher due to latency/throughput limitations.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geonetwork/core-geonetwork/blob/main/CONTRIBUTING.md)
- [x] *Pull request* provided for `main` branch, backports managed with label
- [x] *Good housekeeping* of code, cleaning up comments, tests, and documentation
- [x] *Clean commit history* broken into understandable chucks, avoiding big commits with hundreds of files, cautious of reformatting and whitespace changes
- [x] *Clean commit message*s, longer verbose messages are encouraged
- [ ] *API Changes* are identified in commit messages
- [ ] *Testing* provided for features or enhancements using [automatic tests](https://github.com/geonetwork/core-geonetwork/blob/main/software_development/TESTING.md)
- [ ] *User documentation* provided for new features or enhancements in [manual](https://github.com/geonetwork/core-geonetwork/tree/main/docs/manual)
- [ ] *Build documentation* provided for development instructions in `README.md` files
- [ ] *Library management* using `pom.xml` dependency management. Update build documentation with intended library use and library tutorials or documentation

---
Funded by [LGL BW](https://www.lgl-bw.de/index.html)

tobias-hotz committed:
Handle async responses asynchronously.

This improves indexing time, especially when the ES connection has a high latency or the ES load is high.

tobias-hotz committed:
Delegate the responsibility for batching to the caller

tobias-hotz committed:
Dynamically compute the batch size based on the number of expected index entries

tobias-hotz committed:
Use a cleaner to make sure all documents get send to the index, even in case on of not properly closed submittors

tobias-hotz committed:
Don't use list version of indexMetadata for single entries

tobias-hotz committed:
Remove no longer required method forceIndexChanges

tobias-hotz committed:
Remove commit index changes from frontend

tobias-hotz committed:
Change how a running index job is determined

tobias-hotz committed:
Fixed using the wrong map in BatchingIndexSubmittor

tobias-hotz committed:
Fix field updating not refreshing

This causes issues with some tests. This issue was already present before the changes

tobias-hotz committed:
Fix UserSelectionsApiTest being too strict about the submittor

tobias-hotz committed:
Add support for batch deletion as well

tobias-hotz committed:
Fix batch deletion

tobias-hotz committed:
Allow delete by query to be "batched" by running them async

fxprunayre commented:
Interesting work @tobias-hotz. We are also investigating how to improve indexing performances for GeoNetwork 5. See draft work https://github.com/geonetwork/geonetwork/issues/19 and
https://github.com/geonetwork/geonetwork/blob/main/src/modules/indexing/src/main/java/org/geonetwork/indexing/IndexingService.java#L108
Maybe some ideas can be shared, or tracking some GN4 hot spots to avoid to make similar mistakes in GN5 can be nice.

tobias-hotz commented:
Hi @fxprunayre
thanks for taking a look.
The idea for this is to improve performance of indexing while breaking none of the existing behaviour. Some code paths assume that an element is available in the index right after the call to index, so this has to be taken into account.
Also, the index preparation is still done in the main thread, as it is not unlikely that some code relies on this, and given how big the GN4 Codebase is, I'd rather go with this.

My first approach was to just return the Future of the index response to the caller, but that was getting pretty messy and it was easy to miss a call site. That's why I chose this approach.
For GN5, it could also be benificial to move the index preparation stuff to another thread.

This change allows allows the multithreaded reindexing to work again (which is somewhat broken at the moment, mainly because of concurrency issues with the single document buffer for the bulk requests). This reduces the time spend on reindexing by a lot. So support for multithreaded indexing is something GN5 should also provide out of the box.

tobias-hotz committed:
submittor -> submitter

tobias-hotz committed:
Remove debug sleep

tobias-hotz committed:
Remove unused never implemented that still references forceRefreshReaders

tobias-hotz committed:
Improve log message when deleting

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/geonetwork/core-geonetwork?pullRequest=8465) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/geonetwork/core-geonetwork?pullRequest=8465) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/geonetwork/core-geonetwork?pullRequest=8465) it.</sub>

tobias-hotz committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into async_indexing

# Conflicts:
#	services/src/main/java/org/fao/geonet/api/processing/DatabaseProcessUtils.java

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

tobias-hotz commented:
Hi @fxprunayre,

I hope you're doing well! I wanted to check in about the review of this PR. I completely understand that it’s a large one and might take some time, but it would make a significant difference for us, as well as many other GeoNetwork users.

If you or someone else could find the time to review it, it would be greatly appreciated. If there are any questions or areas you'd like to see adjusted, I’m more than happy to assist with any changes needed. It would be a real shame to see this PR go unnoticed.

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

# Conflicts:
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/oaipmh/OaiPmhHarvester.java
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/thredds/ThreddsHarvester.java
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/webdav/WebDavHarvester.java
#	services/src/main/java/org/fao/geonet/api/records/MetadataWorkflowApi.java
#	web-ui/src/main/resources/catalog/locales/az-v4.json
#	web-ui/src/main/resources/catalog/locales/ca-v4.json
#	web-ui/src/main/resources/catalog/locales/cy-v4.json
#	web-ui/src/main/resources/catalog/locales/da-v4.json
#	web-ui/src/main/resources/catalog/locales/de-v4.json
#	web-ui/src/main/resources/catalog/locales/es-v4.json
#	web-ui/src/main/resources/catalog/locales/fi-v4.json
#	web-ui/src/main/resources/catalog/locales/fr-v4.json
#	web-ui/src/main/resources/catalog/locales/hy-v4.json
#	web-ui/src/main/resources/catalog/locales/is-v4.json
#	web-ui/src/main/resources/catalog/locales/it-v4.json
#	web-ui/src/main/resources/catalog/locales/ka-v4.json
#	web-ui/src/main/resources/catalog/locales/ko-v4.json
#	web-ui/src/main/resources/catalog/locales/nl-v4.json
#	web-ui/src/main/resources/catalog/locales/pt-v4.json
#	web-ui/src/main/resources/catalog/locales/ro-v4.json
#	web-ui/src/main/resources/catalog/locales/ru-v4.json
#	web-ui/src/main/resources/catalog/locales/sk-v4.json
#	web-ui/src/main/resources/catalog/locales/uk-v4.json
#	web-ui/src/main/resources/catalog/locales/zh-v4.json

tobias-hotz committed:
Merge remote-tracking branch 'refs/remotes/upstream/main' into async_indexing

# Conflicts:
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/BaseAligner.java
#	index/src/main/java/org/fao/geonet/index/es/EsRestClient.java
#	web-ui/src/main/resources/catalog/locales/da-v4.json
#	web-ui/src/main/resources/catalog/locales/fr-v4.json

tobias-hotz committed:
Move DatabaseHarvester to new indexing system

tobias-hotz commented:
Hi,

I've updated this PR once again to resolve the merge conflicts and add support to the newly added Database Harvester. I would be really nice if a maintaienr could take a look at this, as it could really help our customer and other users with lots of entries in their catalog.
I know it's a lot to review. If it would help you if we split the indexing and deletion aspect of the PR in two parts, we could also do that.

fxprunayre commented:
Testing the branch, it looks to work well.

tobias-hotz commented:
Thanks a lot for the testing @fxprunayre !
If your testing looks good, can we at least already assign a milestone to this PR so it is not lost among the other Pull Requests?

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/IDeletionSubmitter.java:
```diff
@@ -0,0 +1,12 @@
+package org.fao.geonet.kernel.search.submission;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission;
```

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/IIndexSubmitter.java:
```diff
@@ -0,0 +1,10 @@
+package org.fao.geonet.kernel.search.submission;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission;
```

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingDeletionSubmitter.java:
```diff
@@ -0,0 +1,77 @@
+package org.fao.geonet.kernel.search.submission.batch;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission.batch;
```

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingIndexSubmitter.java:
```diff
@@ -0,0 +1,77 @@
+package org.fao.geonet.kernel.search.submission.batch;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission.batch;
```

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingSubmitterBase.java:
```diff
@@ -0,0 +1,51 @@
+package org.fao.geonet.kernel.search.submission.batch;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission.batch;
```

josegar74 commented on core/src/main/java/org/fao/geonet/kernel/search/submission/batch/StateBase.java:
```diff
@@ -0,0 +1,46 @@
+package org.fao.geonet.kernel.search.submission.batch;
```
```suggestion
/*
 * Copyright (C) 2001-2025 Food and Agriculture Organization of the
 * United Nations (FAO-UN), United Nations World Food Programme (WFP)
 * and United Nations Environment Programme (UNEP)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
 * Rome - Italy. email: geonetwork@osgeo.org
 */
 
package org.fao.geonet.kernel.search.submission.batch;
```

josegar74 committed:
Merge branch 'main' into async_indexing

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/DirectDeletionSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/DirectIndexSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/IDeletionSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/IIndexSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingDeletionSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingIndexSubmitter.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/batch/BatchingSubmitterBase.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Update core/src/main/java/org/fao/geonet/kernel/search/submission/batch/StateBase.java

Co-authored-by: Jose García <josegar74@gmail.com>

tobias-hotz committed:
Fixes after merge

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

# Conflicts:
#	core/src/main/java/org/fao/geonet/kernel/datamanager/IMetadataIndexer.java
#	core/src/main/java/org/fao/geonet/kernel/datamanager/base/BaseMetadataManager.java
#	core/src/main/java/org/fao/geonet/kernel/search/index/BatchOpsMetadataReindexer.java
#	core/src/test/java/org/fao/geonet/AbstractCoreIntegrationTest.java

tobias-hotz commented:
I've rebased this PR onto the latest main branch once more.
Can someone please do a full review so this has a real chance to get in the next release? We've already got some positive test results from our company and @fxprunayre and promising performance numbers.

fxprunayre commented:
Thanks for the additional rebase @tobias-hotz. Indeed the changes looks to work well, no issue reported on indexing on my side. Discussing with @josegar74 yesterday, we propose to make 4.4.9 this month (or early October), and merge that PR just after so that everyone can test it for the 4.4.10 release.

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

# Conflicts:
#	core/src/test/java/org/fao/geonet/AbstractCoreIntegrationTest.java
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/geonet/BaseGeoNetworkAligner.java

tobias-hotz committed:
Fixes after merge

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

# Conflicts:
#	harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/AbstractHarvester.java
#	plugins/datahub-integration/geonetwork-ui

tobias-hotz committed:
Fixes after merge

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

tobias-hotz committed:
Merge remote-tracking branch 'upstream/main' into async_indexing

# Conflicts:
#	core/src/main/java/org/fao/geonet/kernel/search/EsSearchManager.java
#	core/src/main/java/org/fao/geonet/kernel/search/index/BatchOpsMetadataReindexer.java
#	index/src/main/java/org/fao/geonet/index/es/EsRestClient.java
#	plugins/datahub-integration/geonetwork-ui
#	services/src/main/java/org/fao/geonet/api/records/MetadataSharingApi.java
#	services/src/test/java/org/fao/geonet/api/records/MetadataApiTest.java
#	web-ui/src/main/resources/catalog/locales/es-v4.json

tobias-hotz committed:
Fix compile errors after merge

tobias-hotz commented:
Hi, I've updated the PR once more. I've noticed that 4.4.11 was just released, so I gussed that now is probably a good time to attempt to get this merged once more. Please have a look and tell me if you need further changes or if any questions arise.

josegar74 reviewed (commented):
I noticed a difference with previous versions regarding to harvesters progress. Now is updated every 200 records, previously you could see a more fine grained progress. This is not critical, but maybe we can reduce to batches of 50, so the user can see more frequent progress updates.

tobias-hotz commented:
Sure, we can update that value. Note that this PR changes the way the batch size is determined, if expected number of total items is less then 1600, we make smaller batches. See the constructor of [BatchingSubmitterBase](https://github.com/geonetwork/core-geonetwork/pull/8465/changes#diff-f81eb585f86792d635d4ae3e8ad5ff16842a9c81d41391f020bbe5fb1f625481) for reference
Maybe it would be the best if we try to make ~25 instead of 8 steps, and lower the max/default batch size to 100. This would lead to the following batch sizes:
| Estimated Total Size | Batch Size | Total Requests |
| -------------------: | ---------: | -------------: |
|                   50 |         20 |              3 |
|                  200 |         20 |             10 |
|                  500 |         20 |             25 |
|                1,000 |         40 |             25 |
|                2,000 |         80 |             25 |
|                2,500 |        100 |             25 |
|                5,000 |        100 |             50 |
|               10,000 |        100 |            100 |
|               25,000 |        100 |            250 |

I think this might be a good balance between high throughput and the progress not appearing to stall for a long time, but we can also tweak these values further.
We could also introduce config properties for these values, but I am not sure if this is worth it. What do you think?

josegar74 commented:
@tobias-hotz your proposal looks fine to me, thanks. Please update the pull request and we can merge it.

Great work and thanks for your patience rebasing the pull request to keep it up to date.

tobias-hotz committed:
Adjust batch size calculation

tobias-hotz committed:
Trigger another build

tobias-hotz commented:
@josegar74 OK, I've implemented the changes to batch size calculation.

fxprunayre reviewed (approved):
Used in some projects without issue related. Thanks for the improvement.

tobias-hotz commented:
Hi, just checking in: Are we ready to merge this now?
